File: 3331.find-subtree-sizes-after-changes.cpp
#
# @lc app=leetcode id=3331 lang=cpp
#
# [3331] Find Subtree Sizes After Changes
#
# @lc code=start
class Solution {
public:
    vector<int> findSubtreeSizes(vector<int>& parent, string s) {
        int n = parent.size();
        vector<int> newParent = parent;
        
        // For each node from 1 to n-1, find closest ancestor with same character
        for (int i = 1; i < n; i++) {
            char c = s[i];
            int curr = parent[i];
            
            // Travel up the tree to find closest ancestor with same character
            while (curr != -1) {
                if (s[curr] == c) {
                    newParent[i] = curr;
                    break;
                }
                curr = parent[curr];
            }
        }
        
        // Build new tree structure
        vector<vector<int>> children(n);
        for (int i = 1; i < n; i++) {
            children[newParent[i]].push_back(i);
        }
        
        // Calculate subtree sizes using DFS
        vector<int> answer(n);
        function<int(int)> dfs = [&](int node) -> int {
            int size = 1;
            for (int child : children[node]) {
                size += dfs(child);
            }
            return answer[node] = size;
        };
        
        dfs(0);
        return answer;
    }
};
# @lc code=end

File: 3331.find-subtree-sizes-after-changes.py
#
# @lc app=leetcode id=3331 lang=python3
#
# [3331] Find Subtree Sizes After Changes
#

# @lc code=start
from typing import List
class Solution:
    def findSubtreeSizes(self, parent: List[int], s: str) -> List[int]:
        n = len(parent)
        # Step 1: Prepare the tree
        tree = [[] for _ in range(n)]
        for i in range(1, n):
            tree[parent[i]].append(i)

        # Step 2: Find new parent for each node
        new_parent = parent[:]
        # For each character, store stack of (node) in current DFS path
        from collections import defaultdict
        char_stack = defaultdict(list)
        def dfs(u):
            char = s[u]
            char_stack[char].append(u)
            for v in tree[u]:
                # Find closest ancestor with same char
                if len(char_stack[char]) > 1:
                    new_parent[v] = char_stack[char][-2]
                else:
                    new_parent[v] = parent[v]
                dfs(v)
            char_stack[char].pop()
        dfs(0)

        # Step 3: Validate new tree structure
        root_count = sum(1 for p in new_parent if p == -1)
        assert root_count == 1, 'Invalid tree: should have exactly one root.'
        # Build visited set to check connectivity
        visited = set()
        def check(u):
            visited.add(u)
            for v in range(n):
                if new_parent[v] == u:
                    if v in visited:
                        raise Exception('Cycle detected')
                    check(v)
        check(new_parent.index(-1))
        assert len(visited) == n, 'Tree not connected after parent change.'

        # Step 4: Build new tree
        new_tree = [[] for _ in range(n)]
        for i in range(1, n):
            new_tree[new_parent[i]].append(i)

        # Step 5: Compute subtree sizes
        sizes = [0] * n
        def compute(u):
            size = 1
            for v in new_tree[u]:
                size += compute(v)
            sizes[u] = size
            return size
        compute(0)
        return sizes
# @lc code=end